arch: riscv: remove unneeded context switch to gp register

cwshu opened the pull request:
RISC-V global pointer (GP) register is neither caller nor callee register and it's a constant value in the single ELF file.
Thus, we don't need to save/restore GP register at ISR enter/exit. Remove it to optimize context switch performance.

Also, this patch can simply the RISC-V GP support to Zephyr. We could only initialize gp value at program start, w/o initializing it again at each thread init because GP register isn't saved.
I saw #36180 is going to add GP support so this patch will affect it.

cwshu committed:
arch: riscv: fix typo of context switch macro

Fix typo of LOAD_CALLER/CALLEE macros.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu committed:
arch: riscv: remove unneeded context switch to gp register

RISC-V global pointer (GP) register is neither caller nor callee
register, and it's a constant value in the single ELF file. Thus, we
don't need to save/restore GP at ISR enter/exit. Remove it to optimize
context switch performance.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu commented:
Add another reference:
FreeRTOS RISC-V doesn't save GP/TP register because they are constants in it. 
https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/V10.4.4/portable/GCC/RISC-V/portASM.S#L380

Zephyr RISC-V will set TP register to the TLS of thread, but GP register is still a constant so it doesn't need context switch.

cwshu commented:
Sorry for asking, but could anyone help review this patch? It still needs one approval review.
@cfriedt @galak

cfriedt commented:
> Sorry for asking, but could anyone help review this patch? It still needs one approval review.
> @cfriedt @galak

I think I will defer to someone else who has touched this file in the past.

cwshu committed:
Revert "arch: riscv: added support for custom initialization of gp register"

This reverts commit 7b09d031fae3969fbe8647a57190f1fa21fabd8e. Because
context save of GP register is removed, we don't need to initialize GP
at thread init. GP will be a constant value so that it could only be
initialized at program start.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu committed:
soc: riscv: esp32c3: remove RISCV_SOC_INIT_GP_VALUE option.

As this option is not needed and has been removed, also remove it
in the esp32c3 SoC.

Signed-off-by: Jim Shu <cwshu@andestech.com>

cwshu commented:
Revert commit 7b09d031fae3969fbe8647a57190f1fa21fabd8e from #36180 because GP initialization at each thread init is NOT needed after this PR removes the context save/restore of GP register.

Depends on the discussion with @uLipe in #36180, reverting this commit still works and is tested on esp32c3 soc.
